contracts/src/Teleporter/upgrades/TeleporterRegistry.sol

```diff
@@ -0,0 +1,74 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import "../../WarpProtocolRegistry.sol";
+import "../ITeleporterMessenger.sol";
+
+/**
+ * @dev TeleporterRegistry contract is a {WarpProtocolRegistry} and provides an upgrade
+ * mechanism for teleporter messenger contracts.
+ */
+contract TeleporterRegistry is WarpProtocolRegistry {
+    mapping(address => uint256) internal _addressToVersion;
```

Comment:
For my own understanding, what was the motivation to only include the `_addressToVersion` map in the `TeleporterRegistry` rather than putting in the `WarpProtocolRegistery`?

Reply:
Originally was thinking to keep `WarpProtocolRegistry` as abstract as possible, because wasn't sure if other warp protocols would need it. We use it to check `receiveTeleporterMessage` is above a min teleporter version, but not sure if other protocols would use the same thing.

Reply:
I agree with keeping it as abstract as possible. I just think 2 things are little odd:
1. The mapping address to version (or version to address for that matter) isn't specific to Teleporter. It could apply for any protocol really.
2. (Not directly related) `WarpProtocolRegistry` seems like it shouldn't be housed in the `teleporter` repository, if it's meant to be used in a scope broader than Teleporter. Not sure of the best place to put it or other options here, still thinking.

Reply:
What are your thoughts on removing the `WarpProtocolRegistry` abstract contract, and just having `TeleporterRegistry`? My thoughts are:
- Moving `_addressToVersion` logic to `WarpProtocolRegistry` would leave `TeleporterRegistry` with only a few helper functions separately for Teleporter.
- LIke you mentioned, `WarpProtocolRegistry` seems strange to be in Teleporter repo
- Right now we have no plans of building out other warp protocols, so no additional use of `WarpProtocolRegistry`.
- Would simplify some of the code in terms of not needing to override, call base class methods

Reply:
moved `_addressToVersion` logic to `WarpProtocolRegistry`. At this point `TeleporterRegistry` only has helper functions to convert to `ITeleporterMessenger` since dapps will be using the interface for calling it's methods.

Reply:
I agree with all of your thoughts 👍 I thinking unifying them into a single `TeleporterRegistry` implementation at this point makes the most sense.

The only forward-thinking consideration I'm think is if we end up implementing an anycasting protocol as its own contract (say, `TeleporterAnycaster`) or if we ultimately just use warp directly for anycasting. If we were to have a `TeleporterAnycaster` contract, then it'd be nice to be able to use the same registry implementation for it. In that case, it may make sense to have the registry return `address` rather than `ITeleporterMessenger`.

Reply:
hmm with a separate `TeleporterAnycaster` contract I think an option would be to rename `WarpProtocolRegistry` -> `ProtocolRegistry` or `TeleporterRegistry` as base contract, then we'd have separate `TeleporterMessengerRegistry` and `TeleporterAnycasterRegistry`. 

Another option would be to first move everything to `TeleporterRegistry`, then as we develop `TeleporterAnycaster` we refactor as needed, which I don't think is a bad option either.

Reply:
Sorry for the late reply to this thread!

My slight preference is to avoid the "WarpProtocolRegistry" contract name in the `teleporter` repo and add new implementations of it in the future if needed for new Teleporter protocol contracts, but I don't have a strong opinion either way.

Reply:
I am open to this, but if we are moving forward with this, would like it a fast follow separate PR.